### PR TITLE
fix: better conversion error handling in `block_hash_ref`

### PR DESCRIPTION
**Pull request by tcoratger**
Related to https://github.com/paradigmxyz/reth/pull/5864#pullrequestreview-1797075332

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this still panics, no?

File: crates/interfaces/src/provider.rs [hunk 1]
```diff
@@ -122,6 +122,9 @@ pub enum ProviderError {
     /// Snapshot file is not found for requested transaction.
     #[error("not able to find {0} snapshot file for transaction id {1}")]
     MissingSnapshotTx(SnapshotSegment, TxNumber),
+    /// Error during conversion from U256 to u64.
+    #[error("failed to convert U256 to u64")]
+    ConversionError(U256),
```
Comment (mattsse) [inline crates/interfaces/src/provider.rs:127]: this is a bit too unspecific and too specific at the same time.

We want either `BlockNumberOverflow` or a general-purpose Overflow error.

I think the former would be more appropriate

File: crates/revm/src/database.rs [hunk 2]
```diff
@@ -112,9 +112,17 @@ impl<DB: StateProvider> DatabaseRef for StateProviderDatabase<DB> {
...
     /// Returns `Ok` with the block hash if found, or the default hash otherwise.
-    /// Note: This operation may potentially be unsafe due to the unwrap operation.
     fn block_hash_ref(&self, number: U256) -> Result<B256, Self::Error> {
-        // Note: this unwrap is potentially unsafe
-        Ok(self.0.block_hash(number.try_into().unwrap())?.unwrap_or_default())
+        // Attempt to convert U256 to u64
+        let block_number = match number.try_into() {
+            Ok(value) => value,
+            Err(_) => return Err(Self::Error::ConversionError(number)),
+        };
```
Comment (mattsse) [inline crates/revm/src/database.rs:120]: wasn't there another case of try_into?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/interfaces/src/provider.rs (hunk 1 above)
Comment (tcoratger) [reply crates/interfaces/src/provider.rs]: fixed

File: crates/revm/src/database.rs (hunk 2 above)
Comment (tcoratger) [reply crates/revm/src/database.rs]: Just found another one in the same file which is fixed now. I searched the rest of the codebase and couldn't find any others.

File: crates/interfaces/src/provider.rs [hunk 3]
```diff
@@ -122,9 +122,9 @@ pub enum ProviderError {
     /// Snapshot file is not found for requested transaction.
     #[error("not able to find {0} snapshot file for transaction id {1}")]
     MissingSnapshotTx(SnapshotSegment, TxNumber),
-    /// Error during conversion from U256 to u64.
-    #[error("failed to convert U256 to u64")]
-    ConversionError(U256),
+    /// Error encountered when the block number conversion from U256 to u64 causes an overflow.
+    #[error("Failed to convert block number U256 to u64: {0}")]
```
Comment (mattsse) [inline crates/interfaces/src/provider.rs:126]:
```suggestion
    #[error("failed to convert block number U256 to u64: {0}")]
```

**Review by mattsse — APPROVED** [review]
pending @DaniPopes

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**